File: backend-py/src/voxpilot/services/tools/list_directory.py
"""list_directory tool — list entries in a directory."""

from pathlib import Path
from typing import Any, ClassVar

from voxpilot.services.tools.base import Tool

# Directories to always skip when listing
_SKIP_DIRS = {
    ".git",
    "__pycache__",
    "node_modules",
    ".venv",
    "venv",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
}

MAX_ENTRIES = 500


class ListDirectoryTool(Tool):
    """List files and subdirectories in a directory."""

    name = "list_directory"
    description = (
        "List the contents of a directory relative to the working directory. "
        "Returns file and subdirectory names (directories end with '/'). "
        "Common noise directories (.git, __pycache__, node_modules, etc.) are skipped."
    )
    parameters: ClassVar[dict[str, Any]] = {
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": (
                    "Directory path relative to the working directory. "
                    "Defaults to '.' (the working directory itself)."
                ),
            },
        },
        "additionalProperties": False,
    }

    async def execute(self, arguments: dict[str, Any], work_dir: Path) -> str:
        """List a directory's contents."""
        raw_path: str = arguments.get("path", ".")
        resolved = self._resolve_path(raw_path, work_dir)
        if resolved is None:
            return f"Error: path '{raw_path}' is outside the working directory."

        if not resolved.exists():
            return f"Error: directory '{raw_path}' does not exist."
        if not resolved.is_dir():
            return f"Error: '{raw_path}' is not a directory."

        try:
            entries = sorted(resolved.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower()))
        except OSError as exc:
            return f"Error listing '{raw_path}': {exc}"

        lines: list[str] = []
        for entry in entries:
            name = entry.name
            if name in _SKIP_DIRS and entry.is_dir():
                continue
            if entry.is_dir():
                lines.append(f"{name}/")
            else:
                lines.append(name)
            if len(lines) >= MAX_ENTRIES:
                lines.append(f"... (truncated at {MAX_ENTRIES} entries)")
                break

        if not lines:
            return f"Directory '{raw_path}' is empty."

        rel = resolved.relative_to(work_dir.resolve())
        header = f"Directory: {rel}/\n"
        return header + "\n".join(lines)

File: backend/src/tools/list-directory.ts
import type { Dirent } from "node:fs";
import { readdir, stat } from "node:fs/promises";
import { relative, resolve } from "node:path";
import { z } from "zod/v4";
import { resolvePath, simpleResult, type Tool, type ToolResult } from "./base";

const SKIP_DIRS = new Set([
  ".git",
  "__pycache__",
  "node_modules",
  ".venv",
  "venv",
  ".mypy_cache",
  ".pytest_cache",
  ".ruff_cache",
]);

const MAX_ENTRIES = 500;

const parameters = z
  .object({
    path: z
      .string()
      .optional()
      .describe(
        "Directory path relative to the working directory. " +
          "Defaults to '.' (the working directory itself).",
      ),
  })
  .strict();

type Params = z.infer<typeof parameters>;

export class ListDirectoryTool implements Tool<typeof parameters> {
  readonly name = "list_directory";
  readonly description =
    "List the contents of a directory relative to the working directory. " +
    "Returns file and subdirectory names (directories end with '/'). " +
    "Common noise directories (.git, __pycache__, node_modules, etc.) are skipped.";
  readonly parameters = parameters;
  readonly requiresConfirmation = false;

  async execute(args: Params, workDir: string): Promise<ToolResult> {
    const rawPath = args.path && args.path !== "" ? args.path : ".";

    const resolved = await resolvePath(rawPath, workDir);
    if (resolved === null) {
      return simpleResult(
        `Error: path '${rawPath}' is outside the working directory.`,
      );
    }

    let st: Awaited<ReturnType<typeof stat>>;
    try {
      st = await stat(resolved);
    } catch {
      return simpleResult(`Error: directory '${rawPath}' does not exist.`);
    }

    if (!st.isDirectory()) {
      return simpleResult(`Error: '${rawPath}' is not a directory.`);
    }

    let entries: Dirent[];
    try {
      entries = (await readdir(resolved, {
        withFileTypes: true,
        encoding: "utf-8",
      })) as Dirent[];
    } catch (exc) {
      return simpleResult(`Error listing '${rawPath}': ${exc}`);
    }

    // Sort: dirs first, then files, case-insensitive name
    entries.sort((a, b) => {
      const aIsDir = a.isDirectory() ? 0 : 1;
      const bIsDir = b.isDirectory() ? 0 : 1;
      if (aIsDir !== bIsDir) return aIsDir - bIsDir;
      return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
    });

    const lines: string[] = [];
    for (const entry of entries) {
      if (SKIP_DIRS.has(entry.name) && entry.isDirectory()) {
        continue;
      }
      if (entry.isDirectory()) {
        lines.push(`${entry.name}/`);
      } else {
        lines.push(entry.name);
      }
      if (lines.length >= MAX_ENTRIES) {
        lines.push(`... (truncated at ${MAX_ENTRIES} entries)`);
        break;
      }
    }

    if (lines.length === 0) {
      return simpleResult(`Directory '${rawPath}' is empty.`);
    }

    const absWorkDir = resolve(workDir);
    const rel = relative(absWorkDir, resolved);
    const header = `Directory: ${rel || "."}/\n`;
    return simpleResult(header + lines.join("\n"));
  }
}
